src/network/ClientConnectionManager.ts

```diff
@@ -0,0 +1,804 @@
+            });
+    }
+
+    private handleSuccessfulAuth(connection: ClientConnection, response: ClientAuthenticationResponseParams): void {
+        this.checkPartitionCount(response.partitionCount);
+        connection.setConnectedServerVersion(response.serverHazelcastVersion);
+        connection.setRemoteAddress(response.address);
+        connection.setRemoteUuid(response.memberUuid);
+
+        const newClusterId = response.clusterId;
+
+        const initialConnection = this.activeConnections.size === 0;
+        const changedCluster = initialConnection && this.clusterId != null && !newClusterId.equals(this.clusterId);
+        if (changedCluster) {
+            this.logger.warn('ConnectionManager', 'Switching from current cluster: ' + this.clusterId +
+                ' to new cluster: ' + newClusterId);
+            this.client.onClusterRestart();
+        }
+
+        this.activeConnections.set(response.memberUuid.toString(), connection);
+        if (initialConnection) {
+            this.clusterId = newClusterId;
+            if (changedCluster) {
+                this.clientState = ClientState.CONNECTED_TO_CLUSTER;
+                this.initializeClientOnCluster(newClusterId);
+            } else {
+                this.clientState = ClientState.INITIALIZED_ON_CLUSTER;
+                this.emitLifecycleEvent(LifecycleState.CONNECTED);
+            }
+        }
+
+        this.logger.info('ConnectionManager', 'Authenticated with server ' +
+            response.address + ':' + response.memberUuid + ', server version: ' + response.serverHazelcastVersion +
+            ', local address: ' + connection.getLocalAddress());
+        this.emitConnectionAddedEvent(connection);
+
+        // It could happen that this connection is already closed and
+        // onConnectionClose() is called even before the block
+        // above is executed. In this case, now we have a closed but registered
+        // connection. We do a final check here to remove this connection
+        // if needed.
+        if (!connection.isAlive()) {
+            // TODO Can the client be in this state ?
+            this.onConnectionClose(connection);
+        }
+
+    }
+
+    private encodeAuthenticationRequest(): ClientMessage {
+        const clusterName = this.client.getConfig().clusterName;
+        const clientVersion = BuildInfo.getClientVersion();
+        const customCredentials = this.client.getConfig().customCredentials;
+        const clientName = this.client.getName();
+
+        let clientMessage: ClientMessage;
+        if (customCredentials != null) {
+            const credentialsPayload = this.client.getSerializationService().toData(customCredentials).toBuffer();
+
+            clientMessage = ClientAuthenticationCustomCodec.encodeRequest(clusterName, credentialsPayload, this.clientUuid,
+                CLIENT_TYPE, SERIALIZATION_VERSION, clientVersion, clientName, this.labels);
+        } else {
+            // TODO Implement Security config
```

Comment:
It would be great if you could create GH issues to track such TODOs for mandatory features that are not implemented yet.

Reply:
https://github.com/hazelcast/hazelcast-nodejs-client/issues/541